### DISCUSSION TITLE
About the interactive working mechanism between this package and the browser.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [acheong08/ChatGPT#979] [General]
As discussed [here](https://github.com/mmabrouk/chatgpt-wrapper/issues/164#issuecomment-1451303805):

> It looks like it only uses a browser to capture the session cookie, then closes it and just hits the chat.openai.com endpoints directly? if so, and if that works, it's probably a lot more responsive.

I'm not sure if this is true. Any comments and explanations will be appreciated.

Regards,
Zhao

**Comment by acheong08** [comment]
The point of this package is to avoid the use of a browser so that it works better on headless VPS. Authentication is handled by https://github.com/acheong08/OpenAIAuth which is a reverse engineering of their login page. It should be much faster than doing it on the browser. 

As for the how we're calling the API, it's via a closed source Cloudflare bypass for ChatGPT I wrote about a week ago.

**Reply by acheong08** [reply]
`mmabrouk/chatgpt-wrapper` seems like the method we used here about a month ago before Cloudflare bypasses were found

**Reply by hongyi-zhao** [reply]
What do you mean by saying `Cloudflare bypasses`, and is it the current method used by this package?

IMO, many users in the world cannot directly access Cloudflare, so I still don't understand why this approach is being selected here.

BTW, what's the previous method?

**Reply by acheong08** [reply]
> BTW, what's the previous method?

same as `mmabrouk/chatgpt-wrapper`

> What do you mean by saying Cloudflare bypasses, and is it the current method used by this package?

If you go to `chat.openai.com`, you get Cloudflare challenge which requires a browser to solve. The current method gets around that via an unprotected endpoint we found

**Reply by hongyi-zhao** [reply]
Does `an unprotected endpoint` means a potential security risk?

**Reply by acheong08** [reply]
Security risk to them, not us

**Reply by hongyi-zhao** [reply]
If so, they might sooner or later block this authentication method, causing this package to not work.

**Reply by acheong08** [reply]
That's why the bypass server is closed source for now.

As for the authentication, I will always be able to reverse it in a few hours if they change it

**Reply by hongyi-zhao** [reply]
If I understand correctly, this project relies on a private bypass server under the hood to work perfectly.

**Reply by acheong08** [reply]
> If I understand correctly, this project relies on a private bypass server under the hood to work perfectly.

Yes
